draft-ietf-quic-transport.md

```diff
@@ -4807,6 +4807,8 @@ responding to a Retry packet. However, the data sent in these packets could be
 different than what was sent earlier. Sending these new packets with the same
 packet number is likely to compromise the packet protection for those packets
 because the same key and nonce could be used to protect different content.
+A server MAY abort the connection if it detects that the client reset the
```

Comment:
If we're making a normative change, I'd argue for SHOULD

Reply:
MAY or SHOULD abort if detects, but no requirement that it attempt to detect seems like a reasonable compromise.

Reply:
MAY SGTM

Reply:
I don't object to this particular decision, but I would have assumed that you could *always* abort a connection whenever the peer violates a MUST. No?

Reply:
Yes, that is generally true, and this makes it more explicit without adding any requirement on the server.

Reply:
Hmm... So I read this the opposite way, which is that an explicit mention like this implies that there is not a general rule

Reply:
We certainly don't mention a general rule anywhere, but we can't stop an endpoint from enforcing MUSTs on its own. I think it's useful to explicitly call it out out here because this is the sort of place where we want to call explicit attention. 

I would have liked a SHOULD here, but calling it out with a MAY is still useful I think.

Reply:
> We certainly don't mention a general rule anywhere, but we can't stop an endpoint from enforcing MUSTs on its own.

Well, we can't stop them in the "no protocol police" sense, but we certainly can have a MUST-level rule that forbids it.

>  I think it's useful to explicitly call it out out here because this is the sort of place where we want to call explicit attention.

Why?

My general point is that text should do work. And if there is already a general rule permitting to you to error on MUST violations then why does this text need to be there? And working backward from that one might infer that no such rule exists.

Reply:
My argument for why here is because this PN reuse can be a security issue.

That said, you're right that it's not as principled as I would like it to be; there also isn't such a general rule, and you make a good point that it might be useful to have a general rule.

Given that we don't have such a rule at the moment, I would be happy to take this text in and evaluate this with other MAYs in the general sense if there's enough interest to do it. Would you mind filing a separate issue for this?

Reply:
> Given that we don't have such a rule at the moment, I would be happy to take this text in and evaluate this with other MAYs in the general sense if there's enough interest to do it. Would you mind filing a separate issue for this?

I like Jana's suggestion. Can we take this PR and create another issue?

Reply:
> My argument for why here is because this PN reuse can be a security issue.

I disagree with this statement. There's only a security issue if the client resets the 0-RTT packet number space, but the server wouldn't be able to detect that, since it wouldn't even bother to unprotect 0-RTT packets. So the only violation that can be detected here is the reuse of packet numbers in the Initial packet number space. Since the Initial keys are publicly known anyway, this can't be a security issue.

Reply:
Generally speaking, detecting MUST violations is fine for Handshake and ApplicationData packets, after decrypting them, is fine because it will always be about a misbehaving peer, and because we do not (yet) have existing implementations that do not follow the MUSTs.

However, using information that is not protected by AEAD is going to be dangerous, as it is a potential attack vector. For unprotected data (incl. Initial packets), a reader should not assume that a MUST on the sender side implies a "MAY detect" on the receiver side. As can be seen on the issue, we discussed if we might or might not want to have a MAY here, knowing that we have a MUST on the sender side.

To summarize, I tend to believe that having an explicit MAY here makes sense.